[CBRD-26283] shorten the string stored in _db_stored_procedure.target_method when it exeeds column length 1024

hyunikn opened the pull request:
<http://jira.cubrid.org/browse/CBRD-26283>

### Purpose

_db_stored_procedure 의 target_method 컬럼은 SP 의 Java 시그니쳐 (이름, 인자타입, 리턴타입) 을 저장하며, 그 타입은 varchar(1024) 로 정의되어 있다. 이 시그니쳐의 이름과 인자 타입은 SP 실행시에 Java .class 파일에서 target method 를 찾아내는데 쓰인다. (참고: 시그니쳐 안의 return 타입은 현재 쓰이는 곳이 없음) 
SP 의 인자의 갯수가 많을 경우에 시그니쳐의 길이가 1024 길이를 넘어갈 수 있고 이 경우에 Data overflow on data type "character varying" 라는 메시지로 에러가 발생한다. 본 PR 에서 이 문제를 수정한다. 

### Implementation

Java 시그니쳐가 1024 길이를 넘어가는 경우에 저장되는 길이를 줄이기 위해서 시그니쳐 문자열에 다음 변경을 가한다.
- Java 타입 이름에서 java.lang. , java.math. 같은 패키지 이름을 생략, 그리고, PL server 쪽에서 축약된 이름을 인식하도록 구현 추가
- return 타입 지정이 있는 경우 (function) return 키워드를 생략 : 
- 공란 (space character) 를 모두 제거

### Remarks

본 PR 은 11.4 에 대한 것임. 
develop 에 대해서는 target_method 컬럼의 길이 확장과 함께 저장되는 데이터의 일관성을 위해서 본 PR과 동일 변경을 적용 예정. 
싸이트에 이미 배포된 11.4 버전과의 데이터 호환성 문제로 11.4에 대해서는 target_method 컬럼 길이 확장을 적용할 수 없음.

hyunikn committed:
shorten the text stored in _db_stored_procedure.target_method column when it exceeds 1024 characters

hyunikn committed:
remove elements of argClassMap which were added in the previous commit (this should be handled in a separate issue)

hyunikn committed:
add short form of java.lang.Integer[][] and java.lang.Float[][] to argClassMap

kangmin5505 commented:
std::regex 를 사용하여 이 이슈를 해결하는 것보다 회의 때 나온 방법(target_method 배열 크기 증가)으로 수정하는 것이 어떨까요?

현재 로직은 std::regex_replace를 여러 번 호출하면서 불필요한 문자열 복사와 정규식 매칭을 반복하기 때문에 성능상 불리하기도 하지만, 기본적으로 C++ 표준 라이브러리의 <regex> 구현은 다른 정규표현식 라이브러리에 비해 최적화가 부족하여 기본적으로 성능이 좋지 않습니다. 
회의에서 논의된 대로 target_method 배열의 크기를 늘려 이슈를 해결하는 쪽이 단순하면서도 안정적으로 보입니다.

위 의견과 별개로 수정하신 `sp_split_target_signature()`의 몇 가지 이슈는 아래와 같습니다.
1. 인자 부분을 `")a java.lang.string("`으로 작성하여도  정상적으로 java SP 가 생성됨
2. `param_types`의 공백을 지우는 함수는 중간에 있는 공백도 지워서 이상한 결과를 반환함 (ex. `(a java.lang.string)` => `(ajava.lang.string)`

hyunikn commented:
> std::regex 를 사용하여 이 이슈를 해결하는 것보다 회의 때 나온 방법(target_method 배열 크기 증가)으로 수정하는 것이 어떨까요?

위 Remarks 에 언급한 대로, 이미 11.4 가 설치된 곳에 데이터 마이그레이션 없이 이 업그레이드를 적용하기 위해서는 컬럼 사이즈를 변경할 수 없습니다. 
develop 브랜치에 대한 대응은 컬럼 사이즈를 늘리는 것으로 간단이 가능합니다만, 11.4에 있는 데이터에서 마이그레이션 해왔을 때 문제 없이 돌아야 함을 고려해서 동일한 변경을 적용해야 할 것 같습니다.

hyunikn commented:
> 기본적으로 C++ 표준 라이브러리의 구현은 다른 정규표현식 라이브러리에 비해 최적화가 부족하여 기본적으로 성능이 좋지 않습니다. 

CREATE PROCEDURE/FUNCTION 호출은 빈번이 발생하는 편이 아니라서 성능 상 최적화가 크게 필요한 부분은 아닌 것 같습니다. 
그리고, 아주 긴 문자열을 다루는 일도 아니어서 성능 차이는 없다시피 할 것 같습니다. 
그리고, 해당 문자열 축약은 1024 byte 가 넘었을 때만 발생하는 아주 드문 경우임을 고려했을 때 쉽게 사용할 수 있는 라이브러리를 써도 무방하다고 봅니다.

hyunikn commented:
> 1. 인자 부분을 `")a java.lang.string("`으로 작성하여도  정상적으로 java SP 가 생성됨

저장할 때 그 부분의 format 을 체크하지는 않습니다. CREATE PROCEDURE/FUNCTION 은 에러 없이 수행될 것 입니다. 
그런데, 생성된 SP 를 실행할 때 저장된 target_method 컬럼의 값이 PL 서버로 날아가 원하는 자바 메소드를 찾아내게 되는데, 그 때 문제를 일으킬 겁니다. 
PL/CSQL 은 그 부분을 format 에 맞게 생성해 주므로, 저런 문제가 있을리 없고, JSP 의 경우 CREATE PROCEDURE/FUNCTION 문의 name 되에 오는 문자열을 사용자가 적어주게 되는데, 저렇게 틀리게 적어주게 되면 JSP 가 생성은 되나 실행이 안되는 상황이 될 겁니다. 기왕이면 등록할 때 포맷 체크를 해 주면 좋으나, 현재는 안 해주고 있습니다. 개선 사항이네요.

hyunikn commented:
> 2. `param_types`의 공백을 지우는 함수는 중간에 있는 공백도 지워서 이상한 결과를 반환함 (ex. `(a java.lang.string)` => `(ajava.lang.string)`

타입 이름만 적어주도록 사용자 메뉴얼에 나와 있습니다. 인자 이름은 적어 주면 기존에도 오동작 합니다. 
java.lang.String 이 아니라 "a " 까지 포함해서 클래스 이름으로 인식할 겁니다.

kangmin5505 commented:
> > 2. `param_types`의 공백을 지우는 함수는 중간에 있는 공백도 지워서 이상한 결과를 반환함 (ex. `(a java.lang.string)` => `(ajava.lang.string)`
> 
> 타입 이름만 적어주도록 사용자 메뉴얼에 나와 있습니다. 인자 이름은 적어 주면 기존에도 오동작 합니다. java.lang.String 이 아니라 "a " 까지 포함해서 클래스 이름으로 인식할 겁니다.

아래와 같은 경우는 어떻게 되는 건가요?
```
CREATE PROCEDURE test_out(x OUT STRING)
AS LANGUAGE JAVA
NAME 'SpCubrid.outTest(java.lang.String[] o)';
```

hyunikn commented:
> > > 2. `param_types`의 공백을 지우는 함수는 중간에 있는 공백도 지워서 이상한 결과를 반환함 (ex. `(a java.lang.string)` => `(ajava.lang.string)`
> > 
> > 
> > 타입 이름만 적어주도록 사용자 메뉴얼에 나와 있습니다. 인자 이름은 적어 주면 기존에도 오동작 합니다. java.lang.String 이 아니라 "a " 까지 포함해서 클래스 이름으로 인식할 겁니다.
> 
> 아래와 같은 경우는 어떻게 되는 건가요?
> 
> ```
> CREATE PROCEDURE test_out(x OUT STRING)
> AS LANGUAGE JAVA
> NAME 'SpCubrid.outTest(java.lang.String[] o)';
> ```

o 를 빼야 올바로 동작합니다. 타입이름만 적어줘야 합니다. 
저 상태로는 실행할 때 에러가 날 것 입니다.

kangmin5505 commented:
> > > > 2. `param_types`의 공백을 지우는 함수는 중간에 있는 공백도 지워서 이상한 결과를 반환함 (ex. `(a java.lang.string)` => `(ajava.lang.string)`
> > > 
> > > 
> > > 타입 이름만 적어주도록 사용자 메뉴얼에 나와 있습니다. 인자 이름은 적어 주면 기존에도 오동작 합니다. java.lang.String 이 아니라 "a " 까지 포함해서 클래스 이름으로 인식할 겁니다.
> > 
> > 
> > 아래와 같은 경우는 어떻게 되는 건가요?
> > ```
> > CREATE PROCEDURE test_out(x OUT STRING)
> > AS LANGUAGE JAVA
> > NAME 'SpCubrid.outTest(java.lang.String[] o)';
> > ```
> 
> o 를 빼야 올바로 동작합니다. 타입이름만 적어줘야 합니다. 저 상태로는 실행할 때 에러가 날 것 입니다.

그렇군요, 감사합니다. 그렇다면 메뉴얼 수정이 필요할 거 같습니다.
https://www.cubrid.org/manual/internal/ko/develop/pl/jsp.html#java-call-specification

beyondykk9 commented on src/sp/sp_constants.hpp:
```diff
@@ -55,7 +55,7 @@
 #define SP_ATTR_OBJECT_CODE             "ocode"
 
-#define SP_MAX_DEFAULT_VALUE_LEN        255
+#define SP_MAX_DEFAULT_VALUE_LEN        (255)
```
사용하지 않는 매크로인 것 같습니다.

hyunikn replied:
네. 찾아보니 사용하는 곳이 없습니다. 삭제하겠습니다.

beyondykk9 commented on pl_engine/pl_server/src/main/java/com/cubrid/jsp/TargetMethod.java:
```diff
@@ -193,8 +199,18 @@ private static void initArgClassMap() {
+        // why not add other array-array types? TODO
         argClassMap.put("[[Ljava.lang.Integer;", Integer[][].class);
         argClassMap.put("[[Ljava.lang.Float;", Float[][].class);
+
```
argClassMap.put("[[Ljava.lang.Double;", Double[][].class);
...
이런 식으로 나머지 타입에 대해서도 추가가 필요한 건 아닌지요?

hyunikn replied:
이상하게도 기존에 안 되어 있어서 추가를 해 봤더니 SQL TC 한 건에 변경이 생겼습니다. 
별도의 이슈에서 수정하는 것이 어떨까 합니다.

kangmin5505 commented:
> > 기본적으로 C++ 표준 라이브러리의 구현은 다른 정규표현식 라이브러리에 비해 최적화가 부족하여 기본적으로 성능이 좋지 않습니다.
> 
> CREATE PROCEDURE/FUNCTION 호출은 빈번이 발생하는 편이 아니라서 성능 상 최적화가 크게 필요한 부분은 아닌 것 같습니다. 그리고, 아주 긴 문자열을 다루는 일도 아니어서 성능 차이는 없다시피 할 것 같습니다. 그리고, 해당 문자열 축약은 1024 byte 가 넘었을 때만 발생하는 아주 드문 경우임을 고려했을 때 쉽게 사용할 수 있는 라이브러리를 써도 무방하다고 봅니다.

호출이 빈번히 발생하는 편이 아닌 상황이라면, static 변수가 아닌 local 변수로 선언하여 메모리를 낭비하지 않도록 하는 것이 더 적절하지 않을까요?

jongmin-won commented:
> > > > > 2. `param_types`의 공백을 지우는 함수는 중간에 있는 공백도 지워서 이상한 결과를 반환함 (ex. `(a java.lang.string)` => `(ajava.lang.string)`
> > > > 
> > > > 
> > > > 타입 이름만 적어주도록 사용자 메뉴얼에 나와 있습니다. 인자 이름은 적어 주면 기존에도 오동작 합니다. java.lang.String 이 아니라 "a " 까지 포함해서 클래스 이름으로 인식할 겁니다.
> > > 
> > > 
> > > 아래와 같은 경우는 어떻게 되는 건가요?
> > > ```
> > > CREATE PROCEDURE test_out(x OUT STRING)
> > > AS LANGUAGE JAVA
> > > NAME 'SpCubrid.outTest(java.lang.String[] o)';
> > > ```
> > 
> > 
> > o 를 빼야 올바로 동작합니다. 타입이름만 적어줘야 합니다. 저 상태로는 실행할 때 에러가 날 것 입니다.
> 
> 그렇군요, 감사합니다. 그렇다면 메뉴얼 수정이 필요할 거 같습니다. https://www.cubrid.org/manual/internal/ko/develop/pl/jsp.html#java-call-specification

메뉴얼 수정은 필요 없어 보입니다.

메뉴얼에 기재된 SpCubrid.test_out의 Java 메서드 시그니처(즉, AS LANGUAGE JAVA NAME 뒤 따옴표 안의 값)에 공백을 많이 추가해, "공백" 및 "java.lang" 부분이 제거되도록 테스트한 결과, 정상적으로 실행되는 것을 확인했습니다.
- unload/load 이후 해당 SP 정상 수행 확인 완료

```sql
CREATE OR REPLACE PROCEDURE test_out(x OUT STRING)
AS LANGUAGE JAVA
NAME 'SpCubrid.outTest(                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                java.lang.String[] o                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                )';
 
SELECT 'Hi' INTO :out_data FROM db_root;
  'Hi'                
======================
  'Hi'    

CALL test_out(:out_data);
  Result              
======================
  NULL    

SELECT :out_data FROM db_root;
  :out_data           
======================
  'Hello, CUBRID'    

select * from _db_stored_procedure;
<00008> unique_name        : 'dba.test_out'
        sp_name            : 'test_out'
        sp_type            : 1
        return_type        : 0
        arg_count          : 1
        args               : {_db_stored_procedure_args}
        lang               : 1
        pkg_name           : NULL
        is_system_generated: 0
        directive          : 0
        target_class       : 'SpCubrid'
        target_method      : 'outTest(String[]o)'
        owner              : db_user
        comment            : NULL
```

다만, 앞서 예시로 든 (a java.lang.string) ⇒ (ajava.lang.string) 형태의 케이스에 대해서는 조금 더 확인이 필요합니다.
- (a java.lang.String) 형태는 Java 문법상 올바르지 않습니다.
- CUBRID 에서는 이러한 형태가 프로시저 생성 시에는 통과되지만, 실제 실행은 불가능합니다.
- 따라서 이러한 케이스를 고려하면 공백 제거 처리는 문제가 되지 않는다고 생각됩니다.

hyunikn commented:
> 호출이 빈번히 발생하는 편이 아닌 상황이라면, static 변수가 아닌 local 변수로 선언하여 메모리를 낭비하지 않도록 하는 것이 더 적절하지 않을까요?

그렇군요. 수정하겠습니다.

hyunikn commented:
> 메뉴얼 수정은 필요 없어 보입니다.

강민님이 적은 링크를 보시면 메뉴얼에 오동작 할 예제가 있습니다.

hyunikn committed:
. remove an unused macro
. convert several static variables to local variables to save memory consumption

ctshim commented on src/sp/sp_catalog.cpp:
```diff
@@ -969,30 +970,92 @@ void sp_normalize_name (std::string &s)
+      param_types = std::regex_replace (param_types, RE_SPACES, ""); // remove spaces
+
+      // get shorter target_mth by erasing "java.lang." at the start of the type name
+      param_types = std::regex_replace (param_types, RE_COMMA_JAVA_LANG, ",");
+      param_types = std::regex_replace (param_types, RE_PAREN_JAVA_LANG, "(");
+
+      // erasing "java.math." in java.math.BigDecimal
+      param_types = std::regex_replace (param_types, RE_COMMA_JAVA_MATH, ",");
+      param_types = std::regex_replace (param_types, RE_PAREN_JAVA_MATH, "(");
+
+      // erasing "java.sql." in java.sql.{Date, Time, Timestamp}
+      param_types = std::regex_replace (param_types, RE_COMMA_JAVA_SQL, ",");
+      param_types = std::regex_replace (param_types, RE_PAREN_JAVA_SQL, "(");
+
+      // erasing "cubrid.sql." in cubrid.sql.CUBRIDOID
+      param_types = std::regex_replace (param_types, RE_COMMA_CUBRID_SQL, ",");
+      param_types = std::regex_replace (param_types, RE_PAREN_CUBRID_SQL, "(");
```
RE_COMMA_JAVA_LANG, RE_PAREN_JAVA_LANG, RE_COMMA_JAVA_MATH, RE_PAREN_JAVA_MATH, ... 등이 동시에 만족하는 경우도 있을 수 있나요?
그런 경우가 없다면,  `std::regex_search(param_types, RE_COMMA_JAVA_LANG) `를 이용해서 매칭이 되면 그것만 replace하고 나머지는 건너뛰는 방식으로 개선을 할 수 있을 듯 합니다.

hyunikn replied:
동시에 매칭될 수 있습니다.
다만, 네 건을 하나의 RE 로 만들수 있어서 적용했습니다.

hyunikn committed:
simplify regular expressions
